tensorflow/stream_executor/rocm/rocm_dnn.cc

```diff
@@ -592,6 +730,279 @@ class ScopedNormalizeDescriptor {
+    : ScopedFusionPlanBase(parent, miopen_handle, miopenVerticalFusion, input_descriptor) {
+    
+    uint64 hash = GetFusionOpHashValue(input_descriptor, filter_descriptor, conv_descriptor, bias_descriptor, activation_descriptor);
+
+    bool is_compiled = CachedFusionPlans::FindOrCreate(hash, parent, &fusion_plan_, miopenVerticalFusion, input_descriptor);
+    if (!is_compiled) {
+
+      miopenFusionOpDescriptor_t conv_op;
+      miopenStatus_t status = wrap::miopenCreateOpConvForward(parent_, fusion_plan_, &conv_op, conv_descriptor, filter_descriptor);
+      if (status != miopenStatusSuccess) {
+	LOG(FATAL) << "call to miopenCreateOpConvForward failed: " << ToString(status);
+      }
+
+      miopenFusionOpDescriptor_t bias_op;
+      status = wrap::miopenCreateOpBiasForward(parent_, fusion_plan_, &bias_op, bias_descriptor);
+      if (status != miopenStatusSuccess) {
+	LOG(FATAL) << "call to miopenCreateOpBiasForward failed: " << ToString(status);
+      }
+
+      miopenActivationMode_t activation_mode;
+      double alpha, beta, gamma;
+      status = wrap::miopenGetActivationDescriptor(parent_, activation_descriptor, &activation_mode, &alpha, &beta, &gamma);
+      if (status != miopenStatusSuccess) {
+	LOG(FATAL) << "call to miopenGetActivationDescriptor failed: " << ToString(status);
+      }
+
+      miopenFusionOpDescriptor_t actv_op;
+      status = wrap::miopenCreateOpActivationForward(parent_, fusion_plan_, &actv_op, activation_mode);
+      if (status != miopenStatusSuccess) {
+	LOG(FATAL) << "call to miopenCreateOpActivationForward failed: " << ToString(status);
+      }
+
+      status = wrap::miopenCompileFusionPlan(parent_, miopen_handle_, fusion_plan_);
+      if (status != miopenStatusSuccess) {
+	LOG(FATAL) << "call to miopenCompileFusionPlan (CBA) failed: " << ToString(status);
+      }
+    }
+  }
+
+  miopenStatus_t SetConvolutionArgs(const void* filter_data) {
+    float alpha = 1.0;
+    float beta = 0.0;
+    return ScopedFusionPlanBase::SetConvolutionArgs(k_conv_op_idx, &alpha, &beta, filter_data);
+  }
+
+  miopenStatus_t SetBiasArgs(const void* bias_data) {
+    float alpha = 1.0;
+    float beta = 0.0;
+    return ScopedFusionPlanBase::SetBiasArgs(k_bias_op_idx, &alpha, &beta, bias_data);
+  }
+
+  miopenStatus_t SetActivationArgs(miopenActivationDescriptor_t activation_descriptor) {
+    miopenActivationMode_t activ_mode;
+    double activ_alpha, activ_beta, activ_gamma;
+    miopenStatus_t status = wrap::miopenGetActivationDescriptor(parent_, activation_descriptor,	&activ_mode, &activ_alpha, &activ_beta, &activ_gamma);
+    if (status != miopenStatusSuccess) {
+      LOG(FATAL) << "call to miopenGetActivationDescriptor failed: "
+		 << ToString(status);
+    }
+
+    float alpha = 1.0;
+    float beta = 0.0;
+
+    return ScopedFusionPlanBase::SetActivationArgs(k_actv_op_idx, &alpha, &beta, activ_alpha, activ_beta, activ_gamma);
+  }
+
+  uint64 GetFusionOpHashValue(miopenTensorDescriptor_t input_descriptor,
```

Comment:
should this be a member function, or put inside an anonymous namespace just like others?

Reply:
this should be a member function, as each fusion op class should have one. almost made it an abstract virtual function, but then tht would required calling a virtual in a constructor, and that is rightfully frowned upon